azurerm/resource_arm_application_gateway.go

```diff
@@ -0,0 +1,1691 @@
+			"id":   *config.ID,
+			"name": *config.Name,
+			"frontend_ip_configuration_id":   *config.ApplicationGatewayHTTPListenerPropertiesFormat.FrontendIPConfiguration.ID,
+			"frontend_ip_configuration_name": path.Base(*config.ApplicationGatewayHTTPListenerPropertiesFormat.FrontendIPConfiguration.ID),
+			"frontend_port_name":             path.Base(*config.ApplicationGatewayHTTPListenerPropertiesFormat.FrontendPort.ID),
+			"frontend_port_id":               *config.ApplicationGatewayHTTPListenerPropertiesFormat.FrontendPort.ID,
+			"protocol":                       string(config.ApplicationGatewayHTTPListenerPropertiesFormat.Protocol),
+		}
+
+		if config.ApplicationGatewayHTTPListenerPropertiesFormat.HostName != nil {
+			listener["host_name"] = *config.ApplicationGatewayHTTPListenerPropertiesFormat.HostName
+		}
+
+		if config.ApplicationGatewayHTTPListenerPropertiesFormat.SslCertificate != nil {
+			listener["ssl_certificate_name"] = path.Base(*config.ApplicationGatewayHTTPListenerPropertiesFormat.SslCertificate.ID)
+			listener["ssl_certificate_id"] = *config.ApplicationGatewayHTTPListenerPropertiesFormat.SslCertificate.ID
+		}
+
+		if config.ApplicationGatewayHTTPListenerPropertiesFormat.RequireServerNameIndication != nil {
+			listener["require_sni"] = *config.ApplicationGatewayHTTPListenerPropertiesFormat.RequireServerNameIndication
+		}
+
+		result = append(result, listener)
+	}
+
+	return result
+}
+
+func flattenApplicationGatewayProbes(probes *[]network.ApplicationGatewayProbe) []interface{} {
+	result := make([]interface{}, 0, len(*probes))
+
+	for _, config := range *probes {
+		settings := map[string]interface{}{
+			"id":                  *config.ID,
+			"name":                *config.Name,
+			"protocol":            string(config.ApplicationGatewayProbePropertiesFormat.Protocol),
+			"path":                *config.ApplicationGatewayProbePropertiesFormat.Path,
+			"host":                *config.ApplicationGatewayProbePropertiesFormat.Host,
+			"interval":            int(*config.ApplicationGatewayProbePropertiesFormat.Interval),
+			"timeout":             int(*config.ApplicationGatewayProbePropertiesFormat.Timeout),
+			"unhealthy_threshold": int(*config.ApplicationGatewayProbePropertiesFormat.UnhealthyThreshold),
+		}
+
+		result = append(result, settings)
+	}
+
+	return result
+}
+
+func flattenApplicationGatewayRequestRoutingRules(rules *[]network.ApplicationGatewayRequestRoutingRule) []interface{} {
+	result := make([]interface{}, 0, len(*rules))
+
+	for _, config := range *rules {
+		listener := map[string]interface{}{
+			"id":                 *config.ID,
+			"name":               *config.Name,
+			"rule_type":          string(config.ApplicationGatewayRequestRoutingRulePropertiesFormat.RuleType),
+			"http_listener_id":   *config.ApplicationGatewayRequestRoutingRulePropertiesFormat.HTTPListener.ID,
+			"http_listener_name": path.Base(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.HTTPListener.ID),
```

Comment:
when the user specifies that they want an http listener for the routing rules named 'http-listen-foo', Azure creates an id of the form 
```
/subscriptions/<Subscription-Id>/resourceGroups/<Resource-Group-Name>/providers/Microsoft.Network/applicationGateways/<Application-Gateway-Name>/backendHttpSettingsCollection/http-listen-foo
```
So path.Base of the id, will return the name of the listener.
Being the newbie that I am , I stuck a printf of the path.Base execution after this setting and got the following output for a listener named 'gi-agwy727-httplstn':
```
2017/07/27 17:26:24 [DEBUG] plugin: terraform-provider-azurerm: 2017/07/27 17:26:24 [INFO] http listener name: gi-agwy727-httplstn
```